### PR TITLE
habtm sharded-ness fix

### PR DESCRIPTION
**Pull request by osheroff**
@bquorning here's my take on your patch.  

BTW, I had trouble getting a "full" repro going (with .columns failing) in the test environment, as the tests create all tables across all shards, which is tricky to decouple.

### CODE REVIEW

File: test/connection_switching_test.rb [hunk 1]
```diff
@@ -519,11 +519,20 @@ class ShardedModel < ActiveRecord::Base
...
+          if ActiveRecord::VERSION::MAJOR >= 4 && ActiveRecord::VERSION::MINOR >= 2
+            skip("There's a bug with Rails 4.2")
+          end
           a = Account.where(:id => 1001).includes(:people).first
           assert a.people.size > 0
           assert_equal 'slave person', a.people.first.name
         end
 
+        it "sets up has and belongs to many sharded-ness correctly" do
+          if Object.const_defined?('Account::HABTM_People')
```
Comment (bquorning) [inline test/connection_switching_test.rb]: Wouldn't it be better to check on ActiveRecord version here? This test could be randomly skipped and we'd never know.

Comment (osheroff) [inline test/connection_switching_test.rb]: sure, will fix.

File: test/connection_switching_test.rb [hunk 2]
```diff
@@ -519,11 +519,20 @@ class ShardedModel < ActiveRecord::Base
         end
 
         it "will :include things via has_and_belongs associations correctly" do
+          if ActiveRecord::VERSION::MAJOR >= 4 && ActiveRecord::VERSION::MINOR >= 2
+            skip("There's a bug with Rails 4.2")
```
Comment (bquorning) [inline test/connection_switching_test.rb]: It seems that your patch works with AR 4.2 as well – you can remove this `skip`.

File: test/connection_switching_test.rb [hunk 3]
```diff
@@ -519,11 +519,20 @@ class ShardedModel < ActiveRecord::Base
...
+          end
           a = Account.where(:id => 1001).includes(:people).first
           assert a.people.size > 0
           assert_equal 'slave person', a.people.first.name
         end
 
+        it "sets up has and belongs to many sharded-ness correctly" do
+          if ActiveRecord::VERSION::MAJOR >= 4 && ActiveRecord::VERSION::MINOR >= 1
+            assert_equal false, !!Account::HABTM_People.is_sharded?
```
Comment (bquorning) [inline test/connection_switching_test.rb]: Maybe this:
```rb
refute Account::HABTM_People.is_sharded?
```

**Comment by bquorning** [comment]
:+1: This fixes my problem.

**Pull request merged**